### PR TITLE
[CS2113T-T13-2] Nuke

### PR DESCRIPTION
**Pull request by A11riseforme**

### CODE REVIEW

**Comment by A11riseforme** [comment]
@wangqinNick @HAOYUN49 @iceclementi

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -1,34 +1,422 @@
...
+</style>
 
 
-## Product Scope
-### Target user profile
+# **Nuke Developer Guide** <small>v1.5</small>  
 
-{Describe the target user profile}
+By: `CS2113T-T13-2`      Since: `Feb 2020`    
+<small>[Go to Webpage](https://ay1920s2-cs2113t-t13-2.github.io/tp/DeveloperGuide.html)</small>
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md:66]: Good use of links to navigate through the link

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -1,34 +1,422 @@
...
+This feature is facilitated by the `AddModuleCommand` class which add the modules specified by the user. It overrides the `execute()` method which extends from the abstract `AddCommand` class which extends from the abstract `Command` class.   The `execute()` method's role is to execute the adding module operation and do necessary checks .
+
+The `AddCommand` will first try to call the static method `add` in `ModuleManager`  class, which will try to add the module specified by the user, exception will thrown according the following rules:
+
+1. `DuplicateModuleException` will be thrown if the module specified by the user is contained in the `ArrayList` named `moduleList` in `ModuleManager` class.
+2. `ModuleNotProvidedException` will be thrown if the module code specified by the user is not contained in the `HashMap` named `modulesMap` in `ModuleManager` class.
+
+Below are the class-diagram for the involved classes:
+
+![image-20200326014336120](images/Add_Module_Command_Class_Diagram.png)
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: The team have clear and useful class diagram, you may add an interface class for Command class.

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -1,34 +1,422 @@
...
+#### Overview  
+The **delete** feature deletes *modules*, *categories* and *tasks* from the Module, Category and Task List respectively.   
+When the user first requests to execute the **delete** command to delete a directory by providing its name, the application will first filter for directories with matching names. From here, there are **three** possible outcomes:  
+
+1. There are **no** matches --  Nothing is deleted.
+2. There is **one** match -- A prompt will be given to the user to confirm the deletion.
+3. There are **multiple** matches -- The list of matches will be shown to the user, and the user chooses which ones to delete. A further prompt will be given to confirm the deletion(s).
+
+#### Feature Implementation  
+![Delete Command Class Diagram](images/Delete_Command_Class_Diagram.png)
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: I think in my opinion, there are too many detail in one diagram. I suggest we can keep it simple how it is shown in Add :)

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -1,34 +1,422 @@
...
+
+The `DeleteCommand` class in turn extends the `FilterCommand` abstract class. The `FilterCommandClass` contains the following vital methods for filtering:  
+- `createFilteredModuleList()` -- Creates an `ArrayList` of the filtered *modules*.  
+- `createFilteredCategoryist()` -- Creates an `ArrayList` of the filtered *categories*.  
+- `createFilteredTaskList()` -- Creates an `ArrayList`of the filtered *tasks*.  
+
+Lastly, the `FilterCommand` class extends the abstract `Command` class that contains the `execute()` method to execute the actual **delete** command.  
+<br>  
+
+![Prompt Command Class Diagram](images/Prompt_Class_Diagram.png)
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: YES this is perfect amount of information! Clear and minimalist but still provide enough information for the reader to understand the class diagram

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -1,34 +1,422 @@
...
+SUCCESS!! Task(s) have been deleted.
+```
+
+Deletion process ends.  
+
+<br>
+
+Below is a *sequence diagram* to illustrate the above example scenario.  
+
+![Delete Command Sequence Diagram](images/Delete_Command_Sequence_Diagram.png)
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: There is a sequence diagram. 😄 
However are are a few improvement that I can find:
1)  PrepareDeleteTaskCommand is not needed,
2)  createFilterTaskList() instead of just createFilterTaskList
3)  The return arrow if there is no object to return, there is no need for them to label them.

Other then these type of error there is nothing wrong with the sequence diagram !! 
Great use of Sequence diagram to explain how delete function works. Provide enough information
not too much and not too less to the reader to understand the flow of the sequence.
Good job 👍

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -1,34 +1,422 @@
...
+```
+
+1. James will simply enter the command `addm cs3235`
+
+   After the input is parsed as an **add module task** and executed, the `AddModuleCommand#execute()` will call `ModuleManager#add()` to add the module cs3235. In the `ModuleManager#add()` method, it will call `ModuleManager#contains()` to check if the module cs3235 exists in the `ArrayList` named `moduleList` , then it will check if the module code "cs3235" is a key in the `HashMap` named `modulesMap`, after all, it will instantiate an `Module` object with the module code "CS3235" and respective title "Computer Security", then add the object into the `moduleList`.
+
+2. James receive the following feedback:
+
+   ```
+   root :
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: I think that a picture would be better as it show what the actual output should look like 🤔

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -1,34 +1,422 @@
...
+   +--------------------------------------------------------------------------------------------------+
+    NO |  MODULE CODE   |                                 MODULE TITLE
+   +--------------------------------------------------------------------------------------------------+
+    1  |     CS1231     |                             Discrete Structures
+    2  |     CS2100     |                            Computer Organisation
+    3  |     CS2113     |              Software Engineering & Object-Oriented Programming
+    4  |     CS3235     |                              Computer Security
+   +--------------------------------------------------------------------------------------------------+
+   Total modules: 4
+   +--------------------------------------------------------------------------------------------------+
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: The diagram is useful and help the reader understand the output better

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -1,34 +1,422 @@
...
+   addc misc
+   SUCCESS!! Category misc is created.
+   ```
+
+
+
+Below is a *sequence diagram* to illustrate the above example scenario.  
+
+```
+to-do: add the sequence diagram
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: Yes please add more sequence diagram to aid the reader understand the feature more 👍

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -1,34 +1,422 @@
...
+```
+
+Deletion process ends.  
+
+<br>
+
+Below is a *sequence diagram* to illustrate the above example scenario.  
+
+![Delete Command Sequence Diagram](images/Delete_Command_Sequence_Diagram.png)
+<span style="color: green"><small><i>Figure <b>Delete Command Sequence Diagram</b></i></small></span>
```
Comment (NyanWunPaing) [inline docs/DeveloperGuide.md]: I am not sure but I think line crossing through parser eg in sequence diagram 2 is not allowed ? 🤔

**Review by NyanWunPaing — COMMENTED** [review]
Very neat UML diagram. Good job guys!

File: src/main/java/seedu/nuke/command/Command.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
+package seedu.nuke.command;
+
+public abstract class Command {
+    // public static String COMMAND_WORD;
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/Command.java:4]: If it's unnecessary, it may need to be deleted after

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -1,34 +1,422 @@
...
+#### Feature Implementation
+
+This feature is facilitated by the `AddModuleCommand` class which add the modules specified by the user. It overrides the `execute()` method which extends from the abstract `AddCommand` class which extends from the abstract `Command` class.   The `execute()` method's role is to execute the adding module operation and do necessary checks .
+
+The `AddCommand` will first try to call the static method `add` in `ModuleManager`  class, which will try to add the module specified by the user, exception will thrown according the following rules:
+
+1. `DuplicateModuleException` will be thrown if the module specified by the user is contained in the `ArrayList` named `moduleList` in `ModuleManager` class.
+2. `ModuleNotProvidedException` will be thrown if the module code specified by the user is not contained in the `HashMap` named `modulesMap` in `ModuleManager` class.
+
+Below are the class-diagram for the involved classes:
```
Comment (quinnyyy) [inline docs/DeveloperGuide.md]: Could it make sense to add the interaction with ModuleManager to this class diagram?

File: src/main/java/seedu/nuke/command/addcommand/AddCommand.java [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+import seedu.nuke.command.Command;
+
+/**
+ * <h3>Add Command</h3>
+ * A general <b>Command</b> to add a Directory.
+ *
+ * @see Command
+ */
+public abstract class AddCommand extends Command {
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/addcommand/AddCommand.java:11]: Hi there, I see in the figure, there is a 

`getParentDictionary.`

 But I haven't seen it in the code..
image

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -1,34 +1,422 @@
...
+
+Below is a *sequence diagram* to illustrate the above example scenario.  
+
+```
+	// To do Sequence diagram here
+```
+
+
+
+#### 1.2. Add Category and Task Features
```
Comment (quinnyyy) [inline docs/DeveloperGuide.md]: Maybe can add some more visuals to this section? It was a bit hard for me to follow just reading along

File: src/main/java/seedu/nuke/command/addcommand/AddModuleCommand.java [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+import static seedu.nuke.util.ExceptionMessage.MESSAGE_MODULE_NOT_PROVIDED;
+import static seedu.nuke.util.Message.messageAddModuleSuccess;
+
+/**
+ * <h3>Add Module Command</h3>
+ * A <b>Command</b> to add a <b>Module</b> into the <b>Module List</b>.
+ * @see Command
+ * @see Module
+ */
+public class AddModuleCommand extends AddCommand {
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/addcommand/AddModuleCommand.java:26]: There is also without `getParentDictionary()`
image

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (quinnyyy) [reply docs/DeveloperGuide.md]: I agree. maybe it could be cleaned up by removing unimportant attributes in the leaf classes

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -1,34 +1,422 @@
...
+	// To be done.
+```
+
+<br>
+
+## **Setting Up**  
+Refer to the guide [here](#) to set up.  
+
+<br>
+
```
Comment (quinnyyy) [inline docs/DeveloperGuide.md:262]: Maybe adding a high level architecture view at the beginning would be helpful. I think it would've made it easier for me to read the implementation stuff if I had some more context

Comment (quinnyyy) [reply docs/DeveloperGuide.md:262]: i know the developer guide is unfinished but just to keep in mind

File: docs/DeveloperGuide.md [hunk 16]
```diff
@@ -1,34 +1,422 @@
...
+```
+SUCCESS!! Task(s) have been deleted.
+```
+
+Deletion process ends.  
+
+<br>
+
+Below is a *sequence diagram* to illustrate the above example scenario.
```
Comment (quinnyyy) [inline docs/DeveloperGuide.md]: I think the annotations between the sequence diagram and the above example were really well done. It made a complex sequence diagram easy to follow

**Review by quinnyyy — COMMENTED** [review]
Great work overall. I think that there are a few improvements that can be made but overall it was easy to understand

File: src/main/java/seedu/nuke/command/filtercommand/FilterCommand.java [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+            if (taskKeyword.isEmpty()) {
+                taskKeyword = DirectoryTraverser.getBaseTask().getDescription();
+            }
+            return filterTasks(moduleKeyword, categoryKeyword, taskKeyword, isExact);
+        } catch (IncorrectDirectoryLevelException e) {
+            return filterTasks(moduleKeyword, categoryKeyword, taskKeyword, isExact);
+        }
+    }
+
+    protected ArrayList<Directory> createFilteredFileList(String moduleKeyword, String categoryKeyword,
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/filtercommand/FilterCommand.java]: Haven't been written in the Class diagram

image

File: src/main/java/seedu/nuke/command/filtercommand/FilterCommand.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+            if (fileKeyword.isEmpty()) {
+                fileKeyword = DirectoryTraverser.getBaseFile().getFileName();
+            }
+            return filterFiles(moduleKeyword, categoryKeyword, taskKeyword, fileKeyword, isExact);
+        } catch (IncorrectDirectoryLevelException e) {
+            return filterFiles(moduleKeyword, categoryKeyword, taskKeyword, fileKeyword, isExact);
+        }
+    }
+
+    private ArrayList<Directory> filterModules(String moduleKeyword, boolean isExact) {
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/filtercommand/FilterCommand.java]: Is it not important so it wasn't written in the diagrams?
image

File: src/main/java/seedu/nuke/command/filtercommand/deletecommand/DeleteFileCommand.java [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+import static seedu.nuke.util.Message.messageConfirmDeleteFile;
+import static seedu.nuke.util.Message.messagePromptDeleteFileIndices;
+
+/**
+ * <h3>Delete File Command</h3>
+ * A <b>Command</b> to delete <b>Files(s)</b> from the <b>File List</b>.
+ * @see Command
+ * @see TaskFile
+ */
+public class DeleteFileCommand extends DeleteCommand {
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/filtercommand/deletecommand/DeleteFileCommand.java:22]: Haven't been written in the class diagram.. is it not vital?
image

File: src/main/java/seedu/nuke/command/promptcommand/DeleteConfirmationPrompt.java [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+     * Checks if any of the directories to be deleted is the current directory or its parent.
+     *
+     * @param filteredList
+     *  The filtered list of directories to be deleted
+     * @return
+     *  <code>TRUE</code> if the current directory or its parent is being deleted, or <code>FALSE</code> otherwise
+     * @throws IncorrectDirectoryLevelException
+     *  If there is illegal access to a directory
+     */
+    private boolean isCurrentDirectoryInList(ArrayList<Directory> filteredList)
```
Comment (yuchenlichuck) [inline src/main/java/seedu/nuke/command/promptcommand/DeleteConfirmationPrompt.java]: image
Haven't been written in the code. Is it not correct?

**Review by yuchenlichuck — COMMENTED** [review]
Good Code~ You really have written a lot of things and it's very completed. Best Regards

File: docs/DeveloperGuide.md (hunk 16 above)
Comment (DDzuikeai) [inline docs/DeveloperGuide.md]: I think the sequence diagram is really detailed and makes the above scenarios easier to understand. 
But would it be better to place each picture together with its related senario?  It's easier for reader to follow and understand.

File: docs/DeveloperGuide.md (hunk 5 above)
Comment (DDzuikeai) [inline docs/DeveloperGuide.md]: Here are some suggestions based on my understanding:
* 1. Method calls should have a `()`, for example use `executeInitialDelete()` instead of `executeInitialDelete`
* 2. Maybe it's better to add constuct method on the arrow representing the call to the constructor.

File: docs/DeveloperGuide.md [hunk 21]
```diff
@@ -1,34 +1,422 @@
...
+Total categories: 5
++--------------------------------------------------------------------------------------------------+
+```
+
+1. James has two choices:
+
+   1. enter `cd cs3235` to enter the module directory then enter `addc misc` to add the category.
+   2. enter `addc misc -m cs3235` to add the category at the root directory.
+
+   Suppose James use the first method, after the second input, the input is parsed as an **add task** command and executed, the `AddCategoryCommand#execute()` will call ``AddCategoryCommand#getParentDirectory()` to get the current parent directory, then it will instantiate an `Category` object with the name "misc". After which `CategoryManager#add()` will be called to add the new object. In the `CategoryManager#add()` method, it will call `CategoryManager#contains()` method to check if the current parent directory contains the category with name "misc", finally add the object into the `ArrayList` of `categoryList`.
```
Comment (DDzuikeai) [inline docs/DeveloperGuide.md]: Seems that you accidentally typed an extra symbol so the format of this paragraph is incorrect
```
``AddCategoryCommand#getParentDirectory()`
```

File: docs/DeveloperGuide.md [hunk 22]
```diff
@@ -1,34 +1,422 @@
...
+
+1. `ModuleNotFoundException` will be thrown if the specified module code does not exists in the `ArrayList` of Module when adding a new category or task.
+2. `CategoryNotFoundException` will be thrown if the specified category name does not exists in the `ArrayList` of Category in `CategoryManager` in `Module` when adding a new task.
+3. `IncorrectDirectoryLevelException` will be thrown if the user is performing the`addc`/`addt` command in the wrong directory without specifying their full parent directories.(For example, when user trying to execute `addc` in a `Root`/`Task` directory, or trying to execute `addt`  in a `Root`/`Module` directory)
+4. `DuplicateTaskException`/`DuplicateCategoryException` will be thrown if the name of the category/task already exist in the current directory.
+
+Below are the class-diagram for the involved classes:
+
+```
+to-do: add the class-diagram
```
Comment (DDzuikeai) [inline docs/DeveloperGuide.md:645]: The description is quite clear but I still feel a bit hard to follow without a diagram.
Look forward to a detailed diagram!

File: docs/DeveloperGuide.md [hunk 23]
```diff
@@ -1,34 +1,422 @@
...
+
+The add category/task enable the user to add category/task/tag into Category/Task/Tag List.(One to one correspondence, which means the user can only add category to Category List, task to Task List).
+
+When the user first request to execute the **add** command*(assuming the command format given is valid)* to add a category/task by providing its name, the application will parse the input after `addm` /`addc` command word as the name for category/task. From here, there are **three** possible outcomes:
+
+1. The specified upper parent directory does not exist. -- No category/task will be added.
+2. The name of the specified category/task already exist. -- No category/task will be added.
+3. The specified upper parent directory exist and the name of the specified category/task does not exist. -- The specified category/task will be added.
+
+#### Feature Implementation
```
Comment (DDzuikeai) [inline docs/DeveloperGuide.md]: The implementation is quite clear, but have you considered alternative ways to implement this feature?
I think it would be better to make a comparison between your current choice and alternatives you may have considered.
So it's more clear to see pros and cons of your current choice.

File: docs/DeveloperGuide.md [hunk 24]
```diff
@@ -1,34 +1,422 @@
...
+```
+
+<br>
+
+## **Setting Up**  
+Refer to the guide [here](#) to set up.  
+
+<br>
+
+## **Design**
```
Comment (DDzuikeai) [inline docs/DeveloperGuide.md]: I think a `Architecture Diagram` can be provided here. So it's easier to understand the main workflow and how components interact with each other.

**Review by DDzuikeai — COMMENTED** [review]
Really good work! Keep on working and don't forget to add more diagrams!
